rts/Lua/LuaMatrixImpl.cpp

```diff
@@ -0,0 +1,323 @@
+{
+	const TObj* obj = ParseObject<TObj>(objID, L);
+
+	if (obj == nullptr)
+		return false;
+
+	const LocalModelPiece* lmp = ParseObjectConstLocalModelPiece(obj, pieceNum);
+
+	if (lmp == nullptr)
+		return false;
+
+	outMat = lmp->GetModelSpaceMatrix();
+	return true;
+}
+
+template<typename TObj>
+bool LuaMatrixImpl::ObjectPieceMatImpl(const unsigned int objID, const unsigned int pieceNum, bool mult, sol::this_state L)
+{
+	if (!mult)
+		return GetObjectPieceMatImpl<TObj>(objID, pieceNum, mat, L);
+
+	CMatrix44f mat2;
+	if (!GetObjectPieceMatImpl<TObj>(objID, pieceNum, mat2, L))
+		return false;
+
+	mat *= mat2;
+	return true;
+}
+
+void LuaMatrixImpl::ScreenViewMatrix(const sol::optional<bool> mult)
+{
+	AssignOrMultMatImpl(mult, LuaMatrixImpl::VIEWPROJ_MULT_DEFAULT, globalRendering->screenViewMatrix);
+}
+
+void LuaMatrixImpl::ScreenProjMatrix(const sol::optional<bool> mult)
+{
+	AssignOrMultMatImpl(mult, LuaMatrixImpl::VIEWPROJ_MULT_DEFAULT, globalRendering->screenProjMatrix);
+}
+
+void LuaMatrixImpl::ScreenViewProjMatrix(const sol::optional<bool> mult)
+{
+	auto screenViewProjMatrix = (*globalRendering->screenProjMatrix) * (*globalRendering->screenViewMatrix);
+	AssignOrMultMatImpl(mult, LuaMatrixImpl::VIEWPROJ_MULT_DEFAULT, screenViewProjMatrix);
+}
+
+void LuaMatrixImpl::SunViewMatrix(const sol::optional<bool> mult)
+{
+	AssignOrMultMatImpl(mult, LuaMatrixImpl::VIEWPROJ_MULT_DEFAULT, shadowHandler.GetShadowViewMatrix(CShadowHandler::SHADOWMAT_TYPE_DRAWING));
+}
+
+void LuaMatrixImpl::SunProjMatrix(const sol::optional<bool> mult)
+{
+	AssignOrMultMatImpl(mult, LuaMatrixImpl::VIEWPROJ_MULT_DEFAULT, shadowHandler.GetShadowProjMatrix(CShadowHandler::SHADOWMAT_TYPE_DRAWING));
+}
+
+void LuaMatrixImpl::SunViewProjMatrix(const sol::optional<bool> mult)
+{
+	auto viewProj = shadowHandler.GetShadowProjMatrix(CShadowHandler::SHADOWMAT_TYPE_DRAWING) * shadowHandler.GetShadowViewMatrix(CShadowHandler::SHADOWMAT_TYPE_DRAWING);
+	AssignOrMultMatImpl(mult, LuaMatrixImpl::VIEWPROJ_MULT_DEFAULT, viewProj);
+}
+
+void LuaMatrixImpl::Ortho(const float left, const float right, const float bottom, const float top, const float _near, const float _far, const sol::optional<bool> mult)
+{
+	AssignOrMultMatImpl(mult, LuaMatrixImpl::VIEWPROJ_MULT_DEFAULT, CMatrix44f::ClipOrthoProj(left, right, bottom, top, _near, _far, globalRendering->supportClipSpaceControl * 1.0f));
+}
+
+void LuaMatrixImpl::Frustum(const float left, const float right, const float bottom, const float top, const float _near, const float _far, const sol::optional<bool> mult)
+{
+	AssignOrMultMatImpl(mult, LuaMatrixImpl::VIEWPROJ_MULT_DEFAULT, CMatrix44f::ClipPerspProj(left, right, bottom, top, _near, _far, globalRendering->supportClipSpaceControl * 1.0f));
+}
+
+void LuaMatrixImpl::CameraViewMatrix(const sol::optional<unsigned int> cameraIdOpt, const sol::optional<bool> mult)
+{
+	const auto cam = GetCamera(cameraIdOpt);
```

Comment:
here and in other methods:
`auto cam` -> `CCamera* cam`
`auto matIn` -> `CMatrix44f& matIn`